cluster/assemblestate/assemblestate.go

```diff
@@ -0,0 +1,912 @@
+	}
+}
+
+// AuthenticateAndCommit checks that the given [Auth] message is valid and proves
+// knowledge of the shared secret. If this check is passed, we allow mutation of
+// this [AssembleState] via future calls to [AssembleState.VerifyPeer] with the same
+// certificate.
+//
+// An error is returned if the message's HMAC is found to not prove knowledge of
+// the shared secret.
+//
+// This method is to be called by an implementation of the [Transport] interface.
+//
+// This method calls AssembleState.commit with the current state.
+func (as *AssembleState) AuthenticateAndCommit(auth Auth, cert []byte) error {
+	as.lock.Lock()
+	defer as.lock.Unlock()
+
+	fp := CalculateFP(cert)
+
+	expectedHMAC := CalculateHMAC(auth.RDT, fp, as.config.Secret)
+	if !hmac.Equal(expectedHMAC, auth.HMAC) {
+		return errors.New("received invalid HMAC from peer")
+	}
+
+	if _, ok := as.trusted[fp]; ok {
+		if as.trusted[fp].RDT != auth.RDT {
+			return fmt.Errorf("peer %q and %q are using the same TLS certificate", as.trusted[fp].RDT, auth.RDT)
+		}
+	} else {
+		as.trusted[fp] = Peer{
+			RDT:  auth.RDT,
+			Cert: cert,
+		}
+	}
+
+	if existing, ok := as.fingerprints[auth.RDT]; ok {
+		if existing != fp {
+			return fmt.Errorf("peer %q is using a new TLS certificate", auth.RDT)
+		}
+	} else {
+		as.fingerprints[auth.RDT] = fp
+	}
+
+	// check fingerprint consistency if we already have an identity for this peer
+	if id, ok := as.devices.Lookup(auth.RDT); ok && id.FP != fp {
+		return fmt.Errorf("fingerprint mismatch for device %s", auth.RDT)
+	}
+
+	// if we have discovered the route to this peer, we should record an
+	// authoritative route to it. this ensures that we send the route from our
+	// local node to this peer when we publish
+	if addr, ok := as.addresses[fp]; ok {
+		as.selector.AddAuthoritativeRoute(auth.RDT, addr)
+	}
+
+	as.commit(as.export())
+
+	logger.Debugf("got valid auth message from %s", auth.RDT)
+
+	return nil
+}
+
+// VerifyPeer checks if the given certificate is trusted and maps to a known RDT.
+// If it is, then a [PeerHandle] is returned that can be used to modify the state
+// of the cluster on this peer's behalf.
+//
+// An error is returned if the certificate isn't trusted.
+//
+// This method is to be called by an implementation of the [Transport] interface.
+func (as *AssembleState) VerifyPeer(cert []byte) (*PeerHandle, error) {
+	as.lock.Lock()
+	defer as.lock.Unlock()
+
+	fp := CalculateFP(cert)
+
+	p, ok := as.trusted[fp]
+	if !ok {
+		return nil, errors.New("given TLS certificate is not associated with a trusted RDT")
+	}
+
+	return &PeerHandle{
+		as:   as,
+		peer: p.RDT,
+	}, nil
+}
+
+// Run starts the assembly process, managing both the server and periodic client operations.
+// It returns when the context is cancelled, returning the final routes discovered.
+func (as *AssembleState) Run(
+	ctx context.Context,
+	transport Transport,
+	discover Discoverer,
+) (Routes, error) {
```

Comment:
in a local test run there's still quite a bit of this that doesn't seem covered by tests